Amazon Pay ECE: add payment flow using confirmation tokens

annemirasol opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #3737 

## Changes proposed in this Pull Request:
To set the stage for Amazon Pay, we need to introduce a payment flow that uses confirmation tokens instead of manually created payment methods.

Since Amazon Pay does not support `paymentMethodCreation: manual`, we cannot use the same payment element to create all express checkout buttons.

This PR makes the following changes:
- We add a "confirmation token" branch to deferred intent payment processing
- For product page, classic cart, classic checkout and Pay-for-Order pages, we split the express checkout payment types: creating one payment element for each type.
   - This is already our approach for block cart and block checkout. 
   - As we are splitting the express payment types (and as such, we have to mount them in children divs under the main div), the width/layout for the express payment buttons have slightly changed. See comments for the before and after.

## Testing instructions
1. Purchase items using ECE on the different pages that we display them: product page, classic/block checkout, classic/block cart, and Pay-for-Order page. Purchases should behave as expected.
    - As all of our express payment types are set to use manual payment method creation, this is a regression check. Everything should behave the same way as before. 
2. Purchase products using non-express payment methods, e.g. credit card, Klarna. Purchases should behave as expected. 
3. To test the newly-added confirmation tokens flow, uncomment https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3741/files#r1927748266, which will set Google Pay to use the new flow. Do not forget to rebuild your source afterwards.
4. With Google Pay set to use confirmation tokens, purchase items using Google Pay on the different pages where we display them.
6. In wp-admin, check that the orders get set to "Processing" or "Completed". With confirmation tokens, you may have to wait a few minutes (around 3-5 mins) for it. 
7. Test confirmation token flow + manual capture:
    - In Stripe Settings > Transaction Preferences, turn on "capture later".
    - Verify that you can purchase items without any problem.
    - Similar to automatic capture, it may take a few minutes for the order status to change from "Pending payment" to "On hold".
    - Change the status to "Processing" to capture the payment.

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

Commit:
Add payment flow using confirmation tokens

annemirasol commented:
Other tasks:
- [ ] Update e2es

Tested:
- [ ] Purchase from product page
- [ ] Purchase from block cart
- [ ] Purchase from classic cart
- [ ] Purchase from block checkout
- [ ] Purchase from classic cart
- [ ] Purchase from Pay-for-Order page
- [ ] Purchasing Subscriptions
- [ ] Regression check: CC purchases
- [ ] Regression check: Link in the Payment Element
- [ ] Regression check: Link express payment
- [ ] Regression check: refunds
- [ ] Regression check: saved payment methods
- [ ] Handles 3DS flow (e.g. matches Radar rule)

Commit:
Create separate elements for each express checkout type

Commit:
Cleanup and refactor: fix bad merges, add docblocks and comments

Commit:
Fix breaking tests

Commit:
Add unit tests

Commit:
Fix failing JS tests

Commit:
Add JS tests

Commit:
Add Google Pay to confirmation token flow for testing only

annemirasol commented on client/express-checkout/utils/index.js:
```diff
@@ -373,3 +337,23 @@ export const expressCheckoutNoticeDelay = async () => {
+		return false;
+	}
+
+	// TODO: Uncomment me to test Google Pay on confirmation token flow.
+	// Remove me before merging to production.
+	// if ( [ 'googlePay', 'google_pay' ].includes( expressPaymentType ) ) {
+	// 	return false;
+	// }
+
```
👆 Uncomment this section to test the confirmation token flow, using Google Pay.

annemirasol commented:
This PR changes the way we render the express payment buttons in non-block pages: they are now rendered separately to allow for elements with different payment configs.

As such, the way the buttons look have slightly changed. LMK if you see any problems with this, or if you can think of a way to avoid surprising our merchants with a potential layout change.
 
| Before | After |
|--------|-------|
|<img width="800" alt="Screenshot 2025-01-23 at 4 30 40 PM" src="https://github.com/user-attachments/assets/d8adf4fa-7653-4f33-ae54-899e2e1b5024" /> | <img width="800" alt="Screenshot 2025-01-23 at 4 30 10 PM" src="https://github.com/user-attachments/assets/509510d1-fb2a-47e8-a679-152cdb8b0684" /> |
| <img width="640" alt="Screenshot 2025-01-23 at 4 38 16 PM" src="https://github.com/user-attachments/assets/1fe5040e-510d-4fa5-bd35-3e6f2676997b" /> | <img width="640" alt="Screenshot 2025-01-23 at 4 38 48 PM" src="https://github.com/user-attachments/assets/6ff79e66-abca-43f8-9f81-55ba2a2638f0" /> |

Commit:
Fix bad merge

Commit:
Add comments

Commit:
Fix Pay for Order when paying as guest

Commit:
Add support for manual capture and confirmation tokens

Commit:
Merge branch 'develop' into add/3737-confirmation-token-payment-flow

Commit:
Merge branch 'develop' into add/3737-confirmation-token-payment-flow

wjrosa committed:
Merge branch 'develop' into add/3737-confirmation-token-payment-flow

wjrosa commented:
👋 hi! It looks like changelog.txt and readme.txt needs to be updated

wjrosa commented on client/entrypoints/express-checkout/index.js:
```diff
@@ -114,31 +128,56 @@ jQuery( function ( $ ) {
+			const eceButton = wcStripeECE.createButton( elements, {
+				...getExpressCheckoutButtonStyleSettings(),
+				paymentMethods: {
+					amazonPay: 'never',
```
Is this value correct?

annemirasol replied:
For now, yes. We will add the logic that enables it in #3733.

wjrosa commented on client/entrypoints/express-checkout/index.js:
```diff
@@ -114,31 +128,56 @@ jQuery( function ( $ ) {
 
 			const shippingRates = getShippingRates();
```
Maybe this is not needed anymore? Since now we are using `option.shippingRates`. Although below I can see `shippingRates` being passed 🤔 .

annemirasol replied:
This is the part that actually adds it to the `options` object, and passes that object to `createExpressCheckoutElement()` below.

wjrosa commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -2492,16 +2605,37 @@ private function get_existing_compatible_payment_intent( $order, $payment_method
+		// Check if this is for an express payment
+		if ( ! empty( $express_payment_type ) ) {
+			switch ( $express_payment_type ) {
+				case 'link':
```
I think we could use constants here.

annemirasol replied:
- Updated in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3741/commits/8e98e0c800e253ed08e04c7a58fc1d81cc8526c1 to use the already-defined Link constant.
- Added a task to https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3720 to change to constants for express payment types. Literal strings are currently peppered around several PHP and JS files and could use a consolidated cleanup.

wjrosa commented:
> This PR changes the way we render the express payment buttons in non-block pages: they are now rendered separately to allow for elements with different payment configs.
> 
> As such, the way the buttons look have slightly changed. LMK if you see any problems with this, or if you can think of a way to avoid surprising our merchants with a potential layout change.
> 
> Before	After
> <img alt="Screenshot 2025-01-23 at 4 30 40 PM" width="800" src="https://private-user-images.githubusercontent.com/730823/406236614-d8adf4fa-7653-4f33-ae54-899e2e1b5024.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc5ODQxOTMsIm5iZiI6MTczNzk4Mzg5MywicGF0aCI6Ii83MzA4MjMvNDA2MjM2NjE0LWQ4YWRmNGZhLTc2NTMtNGYzMy1hZTU0LTg5OWUyZTFiNTAyNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyN1QxMzE4MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hYWMwYTQxZDcxZDRiYzZlYzRmYTI3NmFiZmJjOWE0MGVmNDJlYTEzZmI1NTZhYzhmNzQ4MWE2ZTM5ODJlMDNiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.PNhl07nuK9_l0xDcvo9YojNUdOaMPPhEqXYZz8r4tiY">	<img alt="Screenshot 2025-01-23 at 4 30 10 PM" width="800" src="https://private-user-images.githubusercontent.com/730823/406237249-509510d1-fb2a-47e8-a679-152cdb8b0684.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc5ODQxOTMsIm5iZiI6MTczNzk4Mzg5MywicGF0aCI6Ii83MzA4MjMvNDA2MjM3MjQ5LTUwOTUxMGQxLWZiMmEtNDdlOC1hNjc5LTE1MmNkYjhiMDY4NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyN1QxMzE4MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZjlhMTkzZjZjNWE4NDIzMjQ2YjNiZjU3MDg5YzkyMjQ4NGJkNTMzNzU4ZjE5NjU0YmYyOWI3MzJkNjYzNThjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.fPJ5zgEg26KFgSy-YTBKimp75zVM5oVEunglsSZLPZs">
> <img alt="Screenshot 2025-01-23 at 4 38 16 PM" width="640" src="https://private-user-images.githubusercontent.com/730823/406237960-1fe5040e-510d-4fa5-bd35-3e6f2676997b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc5ODQxOTMsIm5iZiI6MTczNzk4Mzg5MywicGF0aCI6Ii83MzA4MjMvNDA2MjM3OTYwLTFmZTUwNDBlLTUxMGQtNGZhNS1iZDM1LTNlNmYyNjc2OTk3Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyN1QxMzE4MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YTMzNzVhMmZhMGZhN2M5ZDA2OTgzODQ0ODZlMWZlNDJlOTZkNWVkYTgwMzNiZmVlYWFhY2NiYWQyN2NkMjE2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.UZIgvHZHNrr03MmWy1afWfE5J6cIQZ00QgbL8sJgt-E">	<img alt="Screenshot 2025-01-23 at 4 38 48 PM" width="640" src="https://private-user-images.githubusercontent.com/730823/406238010-6ff79e66-abca-43f8-9f81-55ba2a2638f0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzc5ODQxOTMsIm5iZiI6MTczNzk4Mzg5MywicGF0aCI6Ii83MzA4MjMvNDA2MjM4MDEwLTZmZjc5ZTY2LWFiY2EtNDNmOC05ZjgxLTU1YmEyYTI2MzhmMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEyN1QxMzE4MTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZDRiMjczOTQ3NWU3MzgzZTA3YTlmYWVlNTI5MThkZjQ2OWU0YWJhYjlhMjdkMjQwNDIxYWI1YzhjZmM4OWZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.NU_KIYe1cZVKcbByQMBrQ62LNog9rBGXGqMw0B1f5QY">

I think it is OK for this PR, but we might want to create a new issue to handle this later. Some merchants do complain about these kind of changes.

wjrosa committed:
Merge branch 'develop' into add/3737-confirmation-token-payment-flow

wjrosa reviewed (changes requested):
Nice work, Anne! Code is nice and it mostly works as intended:

> 1. Purchase items using ECE on the different pages that we display them: product page, classic/block checkout, classic/block cart, and Pay-for-Order page. Purchases should behave as expected.

Product page:
<img width="702" alt="Screenshot 2025-01-27 at 10 45 21" src="https://github.com/user-attachments/assets/85f0b5d6-ea8b-4b04-a50b-4c238dd38c4c" />

Classic checkout:
<img width="733" alt="Screenshot 2025-01-27 at 10 53 50" src="https://github.com/user-attachments/assets/8fe14dce-b92f-4bb5-8e55-6a94a4965689" />

Block checkout:
<img width="725" alt="Screenshot 2025-01-27 at 10 57 47" src="https://github.com/user-attachments/assets/7f1fd24f-b3ca-4d04-a160-2e7a3c0e082c" />

Classic cart:
<img width="685" alt="Screenshot 2025-01-27 at 11 02 27" src="https://github.com/user-attachments/assets/751d6e24-51be-41d7-8476-fb85e6ff69a1" />

Blocks cart:
<img width="658" alt="Screenshot 2025-01-27 at 11 01 10" src="https://github.com/user-attachments/assets/dff25dc5-6ebc-4b85-9c60-a8ff7fde9fae" />

Pay for order:
<img width="708" alt="Screenshot 2025-01-27 at 11 05 33" src="https://github.com/user-attachments/assets/bfff9b98-b630-4eab-ae15-0513dee71177" />

> 2. Purchase products using non-express payment methods, e.g. credit card, Klarna. Purchases should behave as expected.
<img width="736" alt="Screenshot 2025-01-27 at 11 10 03" src="https://github.com/user-attachments/assets/a1c17c07-2304-4faf-a047-00e476422838" />

> 4. With Google Pay set to use confirmation tokens, purchase items using Google Pay on the different pages where we display them.
❌  I get this on any page after removing that comment
<img width="1097" alt="Screenshot 2025-01-27 at 11 15 23" src="https://github.com/user-attachments/assets/67af27ca-f380-4d6d-9ffc-a0d461f77375" />

> 5. In wp-admin, check that the orders get set to "Processing" or "Completed". With confirmation tokens, you may have to wait a few minutes (around 3-5 mins) for it.
❌ Not possible to test due the previous error

> 6. Test confirmation token flow + manual capture:
> - Verify that you can purchase items without any problem.
> - Change the status to "Processing" to capture the payment.
❌ Not possible to test due the previous error

Commit:
Fix capture method option for confirmation token flow

Commit:
Feedback: Use Link constant

Commit:
Add changelog and readme entries

annemirasol commented:
> With Google Pay set to use confirmation tokens, purchase items using Google Pay on the different pages where we display them.
❌ I get this on any page after removing that comment
Screenshot 2025-01-27 at 11 15 23

Oof 😓, sorry about that. I added a last-minute tweak in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3741/commits/817749a40885010d564ac75891ad9951e69066fe to support manual captures and ended up breaking automatic capture. Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3741/commits/f124d78b54a8eca80973244cbda2b68a95d22528, and tested for both scenarios!

wjrosa commented:
> > With Google Pay set to use confirmation tokens, purchase items using Google Pay on the different pages where we display them.
> > ❌ I get this on any page after removing that comment
> > Screenshot 2025-01-27 at 11 15 23
> 
> Oof 😓, sorry about that. I added a last-minute tweak in [817749a](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/817749a40885010d564ac75891ad9951e69066fe) to support manual captures and ended up breaking automatic capture. Fixed in [f124d78](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/f124d78b54a8eca80973244cbda2b68a95d22528), and tested for both scenarios!

Works now! Thanks for fixing the issue 👍

Commit:
Remove code for testing Google Pay on the new flow
